csharp/src/Microsoft.ML.OnnxRuntime/OrtValue.shared.cs

```diff
@@ -889,14 +932,196 @@ public static OrtValue CreateMap(OrtValue keys, OrtValue values)
+                return CreateMap(ortValues[0], ortValues[1]);
+            }
+            catch (Exception)
+            {
+                disposableGuard.Dispose();
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Creates a map OrtValue with non-string keys and string values.
+        /// 
+        /// This helps the user not to create OrtValues for keys and values separately.
+        /// The number of elements in keys and values must be the same and they must be in order.
+        /// 
+        /// The OrtValue for keys would be created on top of the managed memory using it directly.
+        /// string values would be converted to UTF-8 encoding and copied to managed memory.
+        /// 
+        /// </summary>
+        /// <typeparam name="K">unmanaged type of keys</typeparam>
+        /// <param name="keys"></param>
+        /// <param name="values">collection of string values</param>
+        /// <returns>Instance of OrtValue</returns>
+        /// <exception cref="ArgumentNullException"></exception>
+        /// <exception cref="ArgumentException"></exception>
+        public static OrtValue CreateMapWithStringValues<K>(K[] keys, IReadOnlyCollection<string> values) where K : unmanaged
+        {
+            if (keys is null || values is null)
+            {
+                throw new ArgumentNullException("Keys or/and values are null");
+            }
+
+            if (keys.Length != values.Count)
+            {
+                throw new ArgumentException("Expecting keys and values same len. " +
+                    $"Received keys: {keys.Length}, Values: {values.Count}");
+            }
+
+            long[] shape = { keys.Length };
+            Span<OrtValue> ortValues = new OrtValue[2];
+            var disposableGuard = new DisposableArray<OrtValue>(ortValues);
+            try
+            {
+                ortValues[0] = CreateTensorValueFromMemory(keys, shape);
+                ortValues[1] = CreateTensorWithEmptyStrings(OrtAllocator.DefaultInstance, shape);
+                int count = 0;
+                foreach (var value in values)
+                {
+                    ortValues[1].FillStringTensorElement(value.AsSpan(), count++);
+                }
+                return CreateMap(ortValues[0], ortValues[1]);
+            }
+            catch (Exception)
+            {
+                disposableGuard.Dispose();
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// A public delegate that will be invoked map keys and values.
+        /// The delegate helps not to deal with the lifespan of intermediate OrtValues.
+        /// </summary>
+        /// <param name="keys">This would always represent a tensor</param>
+        /// <param name="values">Can be any of the Onnx types, but they would all reduce to tensors eventually</param>
+        public delegate void MapVisitor(OrtValue keys, OrtValue values);
```

Comment:
Naively I would have expected this to operate like SequenceElementVisitor and be called once for each key/value pair in the map. Would that be simpler for a user? Not clear to me what the real world use case for this looks like.

Reply:
Will clarify the documentation